src/hdmf/common/table.py

```diff
@@ -1360,42 +1360,64 @@ class DynamicTableRegion(VectorData):
         'table',
     )
 
+    MAX_ROWS_TO_VALIDATE_INIT = int(1e3)
+
     @docval({'name': 'name', 'type': str, 'doc': 'the name of this VectorData'},
             {'name': 'data', 'type': ('array_data', 'data'),
              'doc': 'a dataset where the first dimension is a concatenation of multiple vectors'},
             {'name': 'description', 'type': str, 'doc': 'a description of what this region represents'},
             {'name': 'table', 'type': DynamicTable,
              'doc': 'the DynamicTable this region applies to', 'default': None},
+            {'name': 'validate_data', 'type': bool,
+             'doc': 'whether to validate the data is in bounds of the linked table', 'default': True},
             allow_positional=AllowPositional.WARNING)
     def __init__(self, **kwargs):
-        t = popargs('table', kwargs)
+        t, validate_data = popargs('table', 'validate_data', kwargs)
+        data = getargs('data', kwargs)
+        if validate_data:
+            self._validate_index_in_range(data, t)
+
         super().__init__(**kwargs)
-        self.table = t
+        if t is not None:  # set the table attribute using fields to avoid another validation in the setter
+            self.fields['table'] = t
+
+    def _validate_index_in_range(self, data, table):
+        """If the length of data is small, and if data contains an index that is out of bounds, then raise an error.
+        If the object is being constructed from a file, raise a warning instead to ensure invalid data can still be
+        read.
+        """
+        if table and len(data) <= self.MAX_ROWS_TO_VALIDATE_INIT:
+            for val in data:
+                if val >= len(table) or val < 0:
+                    error_msg = f"DynamicTableRegion index {val} is out of bounds for {type(table)} '{table.name}'."
+                    self._error_on_new_warn_on_construct(error_msg, error_cls=IndexError)
```

Comment:
Instead of iterating through all the values, I think checking for `np.max(data) >= len(table)` would be more efficient.